### PR TITLE
Use the power of adding helper function to simplify code w/ `Mutability`

### PR DESCRIPTION
**Pull request by WaffleLapkin**
r? @compiler-errors

### CODE REVIEW

**Comment by compiler-errors** [comment]
I'm fine with these changes, but I think that `Mutability::choice` (6d53a3bb496ae36b20fb668c3252596ba6795bce) is a bit too clever -- I think it's more readable with an inline `match`.

**Comment by WaffleLapkin** [comment]
@compiler-errors I do think that it's much nicer that way, matching tends to be too verbose IMO. But I can revert that commit.

**Comment by compiler-errors** [comment]
I'd prefer if you did -- since it's only used in a handful of places, I'm not convinced it's as much of an ergonomic improvement compared to the other improvements in this commit stack.

File: compiler/rustc_hir_typeck/src/coercion.rs [hunk 1]
```diff
@@ -108,11 +108,7 @@ fn coerce_mutbls<'tcx>(
     from_mutbl: hir::Mutability,
     to_mutbl: hir::Mutability,
 ) -> RelateResult<'tcx, ()> {
-    match (from_mutbl, to_mutbl) {
-        (hir::Mutability::Mut, hir::Mutability::Mut | hir::Mutability::Not)
-        | (hir::Mutability::Not, hir::Mutability::Not) => Ok(()),
-        (hir::Mutability::Not, hir::Mutability::Mut) => Err(TypeError::Mutability),
-    }
+    if from_mutbl > to_mutbl { Ok(()) } else { Err(TypeError::Mutability) }
```
Comment (compiler-errors) [inline compiler/rustc_hir_typeck/src/coercion.rs:111]: Should this be `>=`? What happens in the `(Not, Not)` case?

Comment (WaffleLapkin) [reply compiler/rustc_hir_typeck/src/coercion.rs]: Yes, already fixed 😅

Comment (WaffleLapkin) [reply compiler/rustc_hir_typeck/src/coercion.rs]: > [00:19] waffle is a compiler: ok lmao
> [00:19] waffle is a compiler: I accidentally restricted coercions to `&mut` -> `&` only
> [00:19] waffle is a compiler: so you can't coerce `&` to `&`

This made it so no function that passes `&` around can compile 👍🏻

File: compiler/rustc_ast_lowering/src/lib.rs [hunk 2]
```diff
@@ -1781,9 +1781,7 @@ impl<'a, 'hir> LoweringContext<'a, 'hir> {
                     // Given we are only considering `ImplicitSelf` types, we needn't consider
                     // the case where we have a mutable pattern to a reference as that would
                     // no longer be an `ImplicitSelf`.
-                    TyKind::Rptr(_, mt)
-                        if mt.ty.kind.is_implicit_self() && mt.mutbl == ast::Mutability::Mut =>
-                    {
+                    TyKind::Rptr(_, mt) if mt.ty.kind.is_implicit_self() && mt.mutbl.is_mut() => {
                         hir::ImplicitSelfKind::MutRef
                     }
                     TyKind::Rptr(_, mt) if mt.ty.kind.is_implicit_self() => {
```
[the remark this answers is not in the document]
Comment (WaffleLapkin) [reply compiler/rustc_ast_lowering/src/lib.rs]: Done

**Pull request merged**